### PR TITLE
If chcon fails, check if label is already correct

### PR DESCRIPTION
**Pull request by rhatdan**
Currently if a user attempts to chcon a file or directory and fails for
any reason check if the file already has the right label, and continue.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@giuseppe @thaJeztah @kolyshkin PTAL

**Comment by rhatdan** [comment]
Attempt to help fix: https://github.com/containers/podman/issues/14786

**Comment by rhatdan** [comment]
[jkroon81](https://github.com/jkroon81) PTAL

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
I think golangci-lint must be updated to a newer version to work with Go 1.18 (we can likely drop older go versions from the CI matrix as well)

**Comment by rhatdan** [comment]
@thaJeztah How do I update lint?

File: go-selinux/rchcon.go [hunk 1]
```diff
@@ -12,7 +12,22 @@ import (
...
 
 func rchcon(fpath, label string) error {
+	slowMode := false
+	// If the current label matches the new label, assume
+	// other labels are correct.
+	if currentLabel, err := lFileLabel(fpath); err == nil {
+		if label == currentLabel {
+			slowMode = true
+		}
+	}
```
Comment (thaJeztah) [inline go-selinux/rchcon.go:22]: Slightly wondering if it's easier to read if it's a one-liner;

```suggestion
	if currentLabel, err := lFileLabel(fpath); err == nil && label == currentLabel {
		slowMode = true
	}
```

File: go.mod [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module github.com/opencontainers/selinux
 
-go 1.13
+go 1.18
 
-require golang.org/x/sys v0.0.0-20191115151921-52ab43148777
+require golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a
```
Comment (thaJeztah) [inline go.mod:5]: Might be good to do this in a separate commit (keeping the "fix" separate from assisting changes)

File: go-selinux/rchcon.go [hunk 3]
```diff
@@ -12,7 +12,22 @@ import (
 )
 
 func rchcon(fpath, label string) error {
+	slowMode := false
```
Comment (thaJeztah) [inline go-selinux/rchcon.go:15]: Perhaps a silly question, but what does "slowMode" mean here? Looking at the changes, it seems that `slowMode` actually enables a "fast path" (check the label, and we're done), but the name of this variable feels like it has the reverse meaning 😅  

Or is reading the label (`lFileLabel()`) slower than setting it (` lSetFileLabel()`) ?

File: go-selinux/rchcon.go (hunk 1 above)
Comment (thaJeztah) [inline go-selinux/rchcon.go:22]: Slightly wondering if it'd be easier to read if this was a one-liner;

```suggestion
	if currentLabel, err := lFileLabel(fpath); err == nil && label == currentLabel {
		slowMode = true
	}
```

File: go-selinux/selinux_linux.go [hunk 4]
```diff
@@ -1102,7 +1102,18 @@ func chcon(fpath string, label string, recurse bool) error {
...
-		return setFileLabel(fpath, label)
+		err := lSetFileLabel(fpath, label)
+		if err == nil {
+			return nil
+		}
+		if errors.Is(err, os.ErrNotExist) {
+			return err
+		}
+		flabel, _ := lFileLabel(fpath)
+		if flabel == label {
```
Comment (thaJeztah) [inline go-selinux/selinux_linux.go:1113]: I _guess_ it's ok(isn), but we don't check the error here, so technically can't assume `flabel` is an ok value

File: go-selinux/selinux_linux.go [hunk 5]
```diff
@@ -1102,7 +1102,18 @@ func chcon(fpath string, label string, recurse bool) error {
...
+			return nil
+		}
+		return err
```
Comment (thaJeztah) [inline go-selinux/selinux_linux.go:1116]: Somewhat related to my earlier question; if reading back the label is faster than setting it, should we reverse this?

Currently, this;

- tries to set the label
- if that fails for _whatever reason_, other than "file doesn't exist", we try to read the existing label
- if that fails, we return the first error

If reading is indeed faster, we could do;


```suggestion
		if flabel, err := lFileLabel(fpath); err == nil && flabel == label {
			// already has the right label
			return nil
		}

		return lSetFileLabel(fpath, label)
```

**Comment by rhatdan** [comment]
@thaJeztah PTAL again.

**Comment by rhatdan** [comment]
@mrunalp @vrothberg  @kolyshkin PTAL
